### PR TITLE
Update CNI plugin to newest version; support ConfigLists

### PR DESCRIPTION
**Pull request by squeed**
**What this PR does / why we need it**: Updates the CNI network plugin to use the newest version of CNI. This brings with it plugin chaining and support for multiple versions.

**Special notes for your reviewer**: This libcni change is backwards-compatible - older plugins will work without any changes needed.

**Release note**:
```release-note
This adds support for CNI ConfigLists, which permit plugin chaining.
```

### CODE REVIEW

**Comment by squeed** [comment]
@freehan

**Comment by feiskyer** [comment]
@squeed Cool. When will CNI v0.5.0 be released?

**Comment by squeed** [comment]
We were hoping to cut a -rc release today or tomorrow, and a full release a few days after that.

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 1]
```diff
@@ -314,10 +315,15 @@ func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kube
 	}
 
 	// Hook container up with our bridge
-	res, err := plugin.addContainerToNetwork(plugin.netConfig, network.DefaultInterfaceName, namespace, name, id)
+	resT, err := plugin.addContainerToNetwork(plugin.netConfig, network.DefaultInterfaceName, namespace, name, id)
 	if err != nil {
 		return err
 	}
+	// Coerce the CNI result version
+	res, err := cnitypes020.GetResult(resT)
```
Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: I assume this is backward compatible, right?

File: pkg/kubelet/network/cni/cni.go [hunk 2]
```diff
@@ -292,5 +318,23 @@ func buildCNIRuntimeConf(podName string, podNs string, podInfraContainerID kubec
...
+	hostPortsParam := make([]map[string]interface{}, len(hostPorts))
+	for _, p := range hostPorts {
+		hostPortsParam = append(hostPortsParam, map[string]interface{}{
+			"host_port":      p.HostPort,
+			"container_port": p.ContainerPort,
+			"protocol":       strings.ToLower(string(p.Protocol)),
+			"host_ip":        p.HostIP,
+		})
+	}
+	// TODO(cdc) CNI #373 needs to be merged first
```
Comment (freehan) [inline pkg/kubelet/network/cni/cni.go]: I assume there will be follow up on this

File: pkg/kubelet/network/cni/cni.go [hunk 3]
```diff
@@ -98,17 +99,40 @@ func getDefaultCNINetwork(pluginDir, binDir, vendorCNIDirPrefix string) (*cniNet
...
+			if err != nil {
+				glog.Warningf("Error loading CNI config file %s: %v", confFile, err)
+				continue
+			}
+			confList, err = libcni.ConfListFromConf(conf)
+			if err != nil {
+				glog.Warningf("Error converting CNI config file %s to list: %v", confFile, err)
+				continue
+			}
+
```
Comment (freehan) [inline pkg/kubelet/network/cni/cni.go]: nit: remove this empty line

File: pkg/kubelet/network/cni/cni.go [hunk 4]
```diff
@@ -98,17 +99,40 @@ func getDefaultCNINetwork(pluginDir, binDir, vendorCNIDirPrefix string) (*cniNet
...
+		var confList *libcni.NetworkConfigList
+		if strings.HasSuffix(confFile, ".conflist") {
+			confList, err = libcni.ConfListFromFile(confFile)
+			if err != nil {
+				glog.Warningf("Error loading CNI config list file %s: %v", confFile, err)
+				continue
+			}
+		} else {
+
```
Comment (freehan) [inline pkg/kubelet/network/cni/cni.go]: nit: remove this empty line

**Review by freehan — COMMENTED** [review]
Look good overall. Waiting for CNI release. Will take a final look later

**Comment by freehan** [comment]
@k8s-bot ok to test

**Comment by freehan** [comment]
Need to run `hack/update-bazel.sh`

File: pkg/kubelet/network/kubenet/kubenet_linux.go (hunk 1 above)
Comment (squeed) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: Correct - this converts the result in to the 0.2.0 format, which is what kubenet already expects.

File: pkg/kubelet/network/cni/cni.go (hunk 2 above)
Comment (squeed) [inline pkg/kubelet/network/cni/cni.go]: Yup :-). It was merged a few hours ago.

**Comment by squeed** [comment]
0.5.0-rc1 was just released. Going to update the PR with that.

**Comment by squeed** [comment]
Ok, all checks are passing. All that's left now is to wait for 0.5.0 to be released - should be sometime this week.

**Comment by squeed** [comment]
CNI v0.5.0 has been updated. This can be considered for merge.

**Comment by squeed** [comment]
Test failures seem like flakes...

**Comment by squeed** [comment]
@freehan ready for review.

**Comment by caseydavenport** [comment]
@squeed @freehan hmm, what's the status of / plan for this guy? Did this not make it into v1.6?

CC @thockin

**Comment by thockin** [comment]
@freehan has been unplanned-out for a couple weeks, and I didn't have
bandwidth to pick up all his PRs.  This feels risky at this point, doesn't
it?  Is it the best fix?  Plugins still have to be updated, right?

**Comment by tomdee** [comment]
@thockin I think without this PR the hostport plugin can't be used.

**Comment by thockin** [comment]
That's fine - kube has its own support for 1.6, plan was to try to
transition in 1.7

**Comment by squeed** [comment]
It would be nice to get this in for 1.6.1, if possible. It does **not** require any plugin changes.

**Comment by squeed** [comment]
@thockin @freehan care to review this? It's ready to go.

**Comment by freehan** [comment]
I am back. Will review this. Thanks!

**Comment by squeed** [comment]
There is a minor bug in cni v0.5.0 - there is a v0.5.2 release coming in a day or so. I will update the PR for that - so DO NOT MERGE for now please.

**Comment by squeed** [comment]
@freehan 0.5.2 has been released and merged in to this PR - ready for review + merge.

**Review by freehan — COMMENTED** [review]
Looks good overall. Just some nits.

**Comment by freehan** [comment]
FYI @dnardo

File: Godeps/Godeps.json [hunk 5]
```diff
@@ -416,18 +416,33 @@
...
 		},
 		{
 			"ImportPath": "github.com/containernetworking/cni/pkg/types",
-			"Comment": "v0.3.0-61-g9d5e6e6",
-			"Rev": "9d5e6e60e79491207834ae8439e80c943db65a69"
+			"Comment": "v0.5.2",
+			"Rev": "137b4975ecab6e1f0c24c1e3c228a50a3cfba75e"
+		},
+		{
+			"ImportPath": "github.com/containernetworking/cni/pkg/types/020",
```
Comment (squeed) [inline Godeps/Godeps.json]: Nope. Godeps is special :-)

File: pkg/kubelet/network/cni/cni.go [hunk 6]
```diff
@@ -294,5 +326,24 @@ func buildCNIRuntimeConf(podName string, podNs string, podInfraContainerID kubec
...
+	portMappingsParam := make([]cniPortMapping, 0, len(portMappings))
+	for _, p := range portMappings {
+		portMappingsParam = append(portMappingsParam, cniPortMapping{
+			HostPort:      p.HostPort,
+			ContainerPort: p.ContainerPort,
+			Protocol:      strings.ToLower(string(p.Protocol)),
+			HostIP:        p.HostIP,
+		})
+	}
+	rt.CapabilityArgs = map[string]interface{}{
```
Comment (squeed) [inline pkg/kubelet/network/cni/cni.go]: That's correct!

File: pkg/kubelet/network/cni/cni.go [hunk 7]
```diff
@@ -243,16 +275,16 @@ func (plugin *cniNetworkPlugin) GetPodNetworkStatus(namespace string, name strin
...
+	rt, err := plugin.buildCNIRuntimeConf(podName, podNamespace, podInfraContainerID, podNetnsPath)
 	if err != nil {
 		glog.Errorf("Error adding network when buliding cni runtime conf: %v", err)
 		return nil, err
 	}
 
 	netconf, cninet := network.NetworkConfig, network.CNIConfig
-	glog.V(4).Infof("About to run with conf.Network.Type=%v", netconf.Network.Type)
-	res, err := cninet.AddNetwork(netconf, rt)
+	glog.V(4).Infof("About to run with conf.Network.Type=%v", netconf.Plugins[0].Network.Type)
```
Comment (squeed) [inline pkg/kubelet/network/cni/cni.go]: Good point (though I kept it as original). I'll log both

**Comment by squeed** [comment]
@freehan updated & rebased, thanks

**Comment by freehan** [comment]
/lgtm
/approve

Thanks!

**Comment by freehan** [comment]
Could you approve this? It needs owners at root.  @thockin

**Comment by squeed** [comment]
Had to rebase. @freehan @thockin is there any chance of getting this reviewed and merged quickly?

File: pkg/kubelet/network/cni/cni.go [hunk 8]
```diff
@@ -96,17 +106,37 @@ func getDefaultCNINetwork(pluginDir, binDir, vendorCNIDirPrefix string) (*cniNet
...
+		}
+		if len(confList.Plugins) == 0 {
+			glog.Warningf("CNI config list %s has no networks, skipping", confFile)
 			continue
 		}
+		confType := confList.Plugins[0].Network.Type
+
 		// Search for vendor-specific plugins as well as default plugins in the CNI codebase.
-		vendorDir := vendorCNIDir(vendorCNIDirPrefix, conf.Network.Type)
+		vendorDir := vendorCNIDir(vendorCNIDirPrefix, confType)
```
Comment (feiskyer) [inline pkg/kubelet/network/cni/cni.go]: Should we search vendor CNI dir for all plugin types, e.g. networks may have different type in `confList.Plugins`

File: pkg/kubelet/network/cni/cni.go (hunk 6 above)
Comment (feiskyer) [inline pkg/kubelet/network/cni/cni.go]: It seems there is no official plugins (e.g. bridge) support port-mappings yet, is it right?

File: pkg/kubelet/network/cni/cni.go (hunk 8 above)
Comment (squeed) [inline pkg/kubelet/network/cni/cni.go]: Maybe, but that is changing behavior. The aim of this PR is only to enable config lists.

Comment (squeed) [inline pkg/kubelet/network/cni/cni.go]: There is a lot of cleaning up that can be done with how k8s uses CNI but this PR is not the place for it, IMHO. Just trying to get the version bump out now.

**Comment by freehan** [comment]
@thockin for approve

**Comment by freehan** [comment]
fyi @dnardo

**Comment by squeed** [comment]
Fingers crossed that this wins the rebase race.

File: pkg/kubelet/network/cni/cni.go (hunk 6 above)
Comment (dcbw) [inline pkg/kubelet/network/cni/cni.go]: @feiskyer we don't intend to put that capability into the main plugins.  Instead, you would create a CNI ConfigList that includes the CNI "portmap" plugin here https://github.com/containernetworking/plugins/pull/1.  When the Kube CNI driver calls AddNetworkList(), libcni will execute each plugin in the list sequentially, passing the results from each plugin along to the next.  The "portmap" plugin consumes that result and sets up the mapings as required.

**Comment by dcbw** [comment]
LGTM

**Comment by redbaron** [comment]
are there any plans to have it in 1.6 branch?

**Comment by bboreham** [comment]
Can I request that the top description of this PR be extended to mention the portmap capability that it adds?

Really I would like it in the 1.7.0 release notes too, but I'm not sure if those can be changed retrospectively.

**Pull request merged**